File: Python/0101. Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#要比較的是left, right subtree是不是可互相翻轉的

class Solution(object):
    def isSymmetric(self, root):
        if not root:
            return True
        return self.compare(root.left, root.right)

    def compare(self, left, right):

        #先排除空節點的情況
        if left == None and right != None: return False
        elif left != None and right == None: return False
        elif left == None and right == None: return True

        #排除空節點後，再排除數值不相同的情況
        elif left.val != right.val: return False

        #剩下: 左右節點不為空、且數值相同的情況
        #此時才做recursion，做下一層的判斷
        outside = self.compare(left.left, right.right) #left subtree:left ; right subtree:right
        inside = self.compare(left.right, right.left)
        isSame = outside and inside #left, right subtree = medium (logic)
        return isSame

File: Java/7. Binary Tree/0101. Symmetric Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isSymmetric(TreeNode root) {
        // 空樹視為對稱
        if (root == null) {
            return true;
        }

        // 檢查左右子樹是否鏡像
        boolean result = compare(root.left, root.right);
        return result;
    }

    private boolean compare(TreeNode left, TreeNode right) {
        // 一邊空、一邊非空 → 不對稱
        if (left == null && right != null) {
            return false;
        }
        if (left != null && right == null) {
            return false;
        }

        // 兩邊都空 → 對稱（這一層結束）
        if (left == null && right == null) {
            return true;
        }

        // 檢查這一層 root 值（這裡就是在檢查各自子樹的 root）
        if (left.val != right.val) {
            return false;
        }

        // 外側：left.left vs right.right
        boolean compareOutside = compare(left.left, right.right);
        if (!compareOutside) {
            return false;
        }

        // 內側：left.right vs right.left
        boolean compareInside = compare(left.right, right.left);
        if (!compareInside) {
            return false;
        }

        // 兩邊都通過才算對稱
        return true;
    }
}
